packages/core/src/vdoc.ts

```diff
@@ -0,0 +1,178 @@
+//
+// Copyright © 2020, 2021 Anticrm Platform Contributors.
+//
+// Licensed under the Eclipse Public License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may
+// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import type { Class, Data, Doc, Emb, Ref } from './classes'
+import core from './component'
+import { Hierarchy } from './hierarchy'
+import { Space } from './security'
+import { Tx, TxProcessor } from './tx'
+import { generateId } from './utils'
+
+/**
+ * Define user versioned document.
+ */
+export interface VDoc extends Doc {
+  _space: Ref<Space>
+  _createdOn: number
+  _createdBy: string
+  _modifiedOn?: number
+  _modifiedBy?: string
+}
+
+export type VDocData<T extends VDoc> = Data<T, VDoc>
+
+/**
+ * General class of transaction for VDoc's.
+ */
+export interface VDocTx extends Tx {
+  objectId: Ref<VDoc>
+  objectSpace: Ref<Space>
+  objectClass: Ref<Class<Doc>>
+
+  _user: string // A user created object
+  _date: number // A time of creation.
+}
+
+/**
+ * Create a VDoc.
+ */
+export interface TxCreateVDoc<T extends VDoc> extends VDocTx {
+  attributes: VDocData<T>
+}
+
+/**
+ * Update VDoc attributes.
+ */
+export interface TxUpdateVDoc<T extends VDoc> extends VDocTx {
+  attributes: Partial<VDocData<T>>
+}
+
+/**
+ * Add item to VDoc embedded collection.
+ *
+ * If localId is missing it will be assigned.
+ */
+export interface TxAddVDocCollection<T extends Emb> extends VDocTx {
```

Comment:
Same thing, should extend `AddCollection` we may discuss it.

Reply:
Done.